src/telemetry/Manager.cc

```diff
@@ -38,6 +53,449 @@ void Manager::InitPostBrokerSetup(broker::endpoint& ep)
+	static auto opts_rt_off_count_bounds = opts_rt->FieldOffset("count_bounds");
+	static auto bounds_default_val = opts_rt->FieldDefault(opts_rt_off_bounds);
+
+	auto bounds_field_val = opts_record->GetField(opts_rt_off_bounds);
+	if ( ! bounds_field_val || bounds_field_val == bounds_default_val )
+		{
+		std::visit(
+			[&opts_record](const auto& histogram_data)
+			{
+				auto bounds_vec = make_intrusive<zeek::VectorVal>(double_vec_type);
+				for ( const auto& b : histogram_data.buckets )
+					bounds_vec->Append(zeek::make_intrusive<zeek::DoubleVal>(b.upper_bound));
+
+				opts_record->Assign(opts_rt_off_bounds, bounds_vec);
+
+				// If this is an int64_t histogram, also fill count_bounds
+				using T = std::decay_t<decltype(histogram_data)>;
+				if constexpr ( std::is_same_v<T, IntHistogramData> )
+					{
+					auto count_bounds_vec = make_intrusive<zeek::VectorVal>(count_vec_type);
+					for ( const auto& b : histogram_data.buckets )
+						count_bounds_vec->Append(val_mgr->Count(b.upper_bound));
+
+					opts_record->Assign(opts_rt_off_count_bounds, count_bounds_vec);
+					}
+			},
+			histogram);
+		}
+
+	return r;
+	}
+
+/**
+ * Encapsulate matching of prefix and name against a broker::telemetry::metric_family_hdl
+ */
+class MetricFamilyMatcher
+	{
+public:
+	MetricFamilyMatcher(std::string_view prefix, std::string_view name)
+		: prefix_pattern(prefix), name_pattern(name)
+		{
+		}
+
+	/**
+	 * @return true if the given family's prefix and name match, else false;
+	 */
+	bool operator()(const broker::telemetry::metric_family_hdl* family)
+		{
+		std::string prefix(broker::telemetry::prefix(family));
+		std::string name(broker::telemetry::name(family));
+
+		if ( fnmatch(prefix_pattern.c_str(), prefix.c_str(), 0 == FNM_NOMATCH) )
+			return false;
+
+		if ( fnmatch(name_pattern.c_str(), name.c_str(), 0) == FNM_NOMATCH )
+			return false;
+
+		return true;
+		}
+
+private:
+	std::string prefix_pattern;
+	std::string name_pattern;
+	};
+
+/**
+ * A collector implementation for counters and gauges.
+ */
+class MetricsCollector : public broker::telemetry::metrics_collector
+	{
+	using MetricType = Manager::MetricType;
+
+public:
+	MetricsCollector(std::string_view prefix, std::string_view name) : matches(prefix, name) { }
+
+	void operator()(const broker::telemetry::dbl_counter_family_hdl* family,
+	                const broker::telemetry::dbl_counter_hdl* counter,
+	                broker::telemetry::label_list labels)
+		{
+		if ( matches(upcast(family)) )
+			metrics.emplace_back(Manager::CollectedValueMetric::NewCounter(
+				broker::telemetry::upcast(family), extract_label_values(labels),
+				broker::telemetry::value(counter)));
+		}
+
+	void operator()(const broker::telemetry::int_counter_family_hdl* family,
+	                const broker::telemetry::int_counter_hdl* counter,
+	                broker::telemetry::label_list labels)
+		{
+		if ( matches(upcast(family)) )
```

Comment:
Since all overloads `upcast` the family object, why not consider just passing `metric_family_hdl*` in from Broker in the first place? (see: https://github.com/zeek/broker/pull/258#discussion_r934782170).